File: packages/sandboxd/src/protocol/runtime_state.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RuntimeAttachment {
    pub sandbox_instance_id: String,
    pub owner_lease_id: String,
    pub node_id: String,
    pub session_id: String,
    pub attached_at_ms: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RuntimePresenceSummary {
    pub active_count: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RuntimeKeepaliveSummary {
    pub active: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RuntimeStateSnapshot {
    pub owner_lease_id: Option<String>,
    pub attachment: Option<RuntimeAttachment>,
    pub presence: RuntimePresenceSummary,
    pub keepalive: RuntimeKeepaliveSummary,
}

File: packages/sandboxd/protocol/runtime_state.go
package protocol

type RuntimeAttachment struct {
	SandboxInstanceID string `json:"sandboxInstanceId"`
	OwnerLeaseID      string `json:"ownerLeaseId"`
	NodeID            string `json:"nodeId"`
	SessionID         string `json:"sessionId"`
	AttachedAtMS      uint64 `json:"attachedAtMs"`
}

type RuntimePresenceSummary struct {
	ActiveCount uint64 `json:"activeCount"`
}

type RuntimeKeepaliveSummary struct {
	Active bool `json:"active"`
}

type RuntimeSummary struct {
	Ready bool `json:"ready"`
}

type RuntimeStateSnapshot struct {
	OwnerLeaseID *string                 `json:"ownerLeaseId"`
	Attachment   *RuntimeAttachment      `json:"attachment"`
	Presence     RuntimePresenceSummary  `json:"presence"`
	Keepalive    RuntimeKeepaliveSummary `json:"keepalive"`
	Runtime      RuntimeSummary          `json:"runtime"`
}

func DecodeRuntimeStateSnapshot(data []byte) (RuntimeStateSnapshot, error) {
	var snapshot RuntimeStateSnapshot
	if err := decodeStrict(data, &snapshot); err != nil {
		return RuntimeStateSnapshot{}, err
	}
	return snapshot, nil
}
